### PR TITLE
std.algorithm.search: expose extremum

### PR DESCRIPTION
**Pull request by wilzbach**
On #4221 the consensus was that extremum is rarely needed, hence we set it to `private`.

It seems like @9il just found a nice use case for extremum, `std.math.cmp`:

```
assert([-2.0, 2, 4].reduce!`cmp(a, b) ? a : b` == -2.0);
assert([-2.0, 2, 4].reduce!`cmp(a, b) ? b : a` == 4.0);
```

Well it's also possible to use `reduce`:

```
assert([-1., 0, 1].extremum!cmp == 1.0);
assert([-1., 0, 1].extremum!`cmp(a, b) < 0` == -1.0);
```

However it gets neat, if map is needed:

```
assert([-1., 0, 1].enumerate.extremum!cmp == tuple(2, 1.0));
assert([-1., 0, 1].enumerate.extremum!`cmp(a, b) < 0` == tuple(0, -1.0));
```

I added a small goodie that checks whether the first template function can be used as an mapping function (unary function) and otherwise it falls back to the selector (binary function).

Additionally I added more tests.

### CODE REVIEW

**Comment by wilzbach** [comment]
Btw during the minElement review no one complained, but is there a way to avoid the mapping element if a "a" noop function is used?

**Comment by 9il** [comment]
ping @DmitryOlshansky

**Comment by wilzbach** [comment]
> Btw during the minElement review no one complained, but is there a way to avoid the mapping element if a "a" noop function is used?

Yeah got it too work, so that now a `IdentityFun` is passed as default argument from `extremum` and `{min,max}Element` and statically checked against in the `extremum` function :)

**Comment by wilzbach** [comment]
> Yeah got it too work, so that now a IdentityFun is passed as default argument from extremum and {min,max}Element and statically checked against in the extremum function

Should I separate this optimization? (it doesn't depend on extremum being exposed to the public)

**Comment by 9il** [comment]
What optimization and why you think that this is optimization?

**Comment by wilzbach** [comment]
> What optimization and why do you think that this is optimization?

If  the alias is "a" (e.g. [0,2,3].minElement), currently this is executed:

```
MapType mapElement = mapFun(r[i]);
if (selectorFun(mapElement, extremeElementMapped))
{
   extremeElement = r[i];
   extremeElementMapped = mapElement;
}
```

this PR reduces it down to:

```
if (selectorFun(r[i], extremeElement))
{
   extremeElement = r[i];
}
```

(have a look at the changes)

**Comment by wilzbach** [comment]
> Should I separate this optimization? (it doesn't depend on extremum being exposed to the public)

See #4265 ;-)

**Comment by 9il** [comment]
> If the alias is "a" (e.g. [0,2,3].minElement), currently this is executed:
> 
> MapType mapElement = mapFun(r[i]);
> if (selectorFun(mapElement, extremeElementMapped))
> {
>    extremeElement = r[i];
>    extremeElementMapped = mapElement;
> }
> this PR reduces it down to:
> 
> if (selectorFun(r[i], extremeElement))
> {
>    extremeElement = r[i];
> }
> (have a look at the changes)

they are equivalent, are not they? (i mean assembler code)

**Comment by wilzbach** [comment]
> they are equivalent, are not they? (i mean assembler code)

Have a look: http://www.mergely.com/966byvI1/ - (e.g. unaryFun is called on the right)

**Comment by 9il** [comment]
> they are equivalent, are not they? (i mean assembler code)
> Have a look: http://www.mergely.com/966byvI1/ - (e.g. unaryFun is called on the right)
>  Hide all checks

Try with `ldc2 -O -release -boundscheck=off -output-s`

**Comment by andralex** [comment]
Can't we define overloads to avoid map if not specified? Should we promote only to `package` level?

**Comment by wilzbach** [comment]
> Can't we define overloads to avoid map if not specified?

See #4265.

> Should we promote only to package level?

The code has been moved to #4265 - this is just about whether `extremum` should be publicly exposed.
See my first post for an example for which `extremum` could be useful.

File: std/algorithm/searching.d [hunk 1]
```diff
@@ -1239,7 +1241,7 @@ body
     return extremum!(map, selector)(r, seed);
 }
 
-private auto extremum(alias map = "a", alias selector = "a < b", Range,
+auto extremum(alias map = "a", alias selector = "a < b", Range,
```
Comment (andralex) [inline std/algorithm/searching.d]: `///` before this?

**Comment by wilzbach** [comment]
>  this is just about whether `extremum` should be publicly exposed.
> Please see my first post for an example for which `extremum` could be useful.

Rebased to latest master. So do we want to expose `extremum`?

**Comment by andralex** [comment]
I approve the addition, seeing I have a soft spot for `extremum` too. @wilzbach could you please detect the identity mapping via a specialization (as opposed to string comparison)? Seems more elegant. LMK.

**Comment by wilzbach** [comment]
> andralex commented on Dec 24, 2016
> I approve the addition, seeing I have a soft spot for extremum too. 

While it would have been a nice Christmas present, after thinking about this for a while I reached my internal consensus that the use cases of a public `extremum` would be pretty low and exposing it to the general public would lead to more confusion than help.
Using `reduce` works just fine and if real use cases where it would help to simplify code still arise, we can always revert this decision, but once exposed there's no going back.

> @wilzbach could you please detect the identity mapping via a specialization (as opposed to string comparison)? Seems more elegant. LMK.

With #5001 we finally got this in :)

**Pull request closed without merging**